File: leetcode-go/Q051-Q100/Q063_Unique_Paths_II/Q063_Unique_Paths_II.go
package Q063_Unique_Paths_II

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	m, n := len(obstacleGrid), len(obstacleGrid[0])

	dp := make([][]int, m)
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n)
	}
	for i := 0; i < m; i++ {
		if obstacleGrid[i][0] == 1 {
			break
		}
		dp[i][0] = 1
	}
	for j := 0; j < n; j++ {
		if obstacleGrid[0][j] == 1 {
			break
		}
		dp[0][j] = 1
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if obstacleGrid[i][j] == 1 {
				continue
			}
			if obstacleGrid[i-1][j] == 0 {
				dp[i][j] += dp[i-1][j]
			}
			if obstacleGrid[i][j-1] == 0 {
				dp[i][j] += dp[i][j-1]
			}
		}
	}
	return dp[m-1][n-1]
}

File: leetcode-java/src/dp/Q063_Unique_Paths_II.java
package dp;

public class Q063_Unique_Paths_II {

    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length, n = obstacleGrid[0].length;
        int[][] dp = new int[m][n];
        dp[0][0] = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (obstacleGrid[i][j] == 1) {
                    dp[i][j] = 0;
                    continue;
                }
                if (i > 0 && dp[i-1][j] != 0) dp[i][j] += dp[i-1][j];
                if (j > 0 && dp[i][j-1] != 0) dp[i][j] += dp[i][j-1];
            }
        }
        return dp[m-1][n-1];
    }

}
